### DISCUSSION TITLE
Version: undefined   and   Build time: undefined

### DISCUSSION THREAD

**Discussion by tonho911** [0xERR0R/blocky#1286] [Q&A]
morning sir...
when I finish building blocky for mipsle softfloat 
![Screenshot_97](https://github.com/0xERR0R/blocky/assets/148846306/c1d58ae6-ca34-4e7d-a736-52fb285d79a6)
where can I edit the build version and build time while running blocky?

**Comment by kwitsch** [comment]
It's part of the [Makefile](https://github.com/0xERR0R/blocky/blob/main/Makefile#L19).
You could define these values during build time with the LD compiler flags.

**Reply by tonho911** [reply]
I used this command "GOOS=linux GOARCH=mipsle GOMIPS=softfloat CGO_ENABLED=0 go build -ldflags "-s -w" " where is it wrong bro?

**Reply by kwitsch** [reply]
If you don't use the Makefile you need to copy the line I linked in my first post to your command to set the variable during build time.

**Reply by tonho911** [reply]
edit this sir?
VERSION?=$(shell git describe --always --tags)
BUILD_TIME?=$(shell date --iso-8601=seconds)

**Reply by kwitsch** [reply]
Something like this should work:
```Bash
VERSION=$(shell git describe --always --tags) BUILD_TIME=$(shell date --iso-8601=seconds) GOOS=linux GOARCH=mipsle GOMIPS=softfloat CGO_ENABLED=0 go build -ldflags "-s -w -X github.com/0xERR0R/blocky/util.Version=${VERSION} -X github.com/0xERR0R/blocky/util.BuildTime=$(shell date -d "${BUILD_TIME}" '+%Y%m%d-%H%M%S') -X github.com/0xERR0R/blocky/util.Architecture=${GOARCH}"
```

**Reply by ThinkChaos** [reply]
That's unfortunately not going to work as is: env vars you set at the beginning of a command are set just before executing the command and after var expansion. Example: `a=b echo $a` only prints a newline.

I think the easiest is to do:
`make build GOOS=linux GOARCH=mipsle GOMIPS=softfloat GO_SKIP_GENERATE=1`

`GO_SKIP_GENERATE` is needed because since we set cross compile Go vars, `go generate` will use the `mipsle` Go and crash.  
I guess it would be nice to remove the `build: generate` dependency so this is more straightforward. We can still have it be done as part of `all`

**Reply by kwitsch** [reply]
If you don't use the Makefile but build it without it the provided example should work.

As far as I got it the binary was build without `make`.

**Reply by ThinkChaos** [reply]
No it won't because of the vars issue mentioned above and the `$(shell ...)` pattern.
You'd need to do:
```sh
# First set the vars we need in the CLI args
VERSION=$(git describe --always --tags)
BUILD_TIME=$(date --iso-8601=seconds)
export GOARCH=mipsle

# Then run the command
GOOS=linux GOMIPS=softfloat CGO_ENABLED=0 go build -ldflags "-s -w -X github.com/0xERR0R/blocky/util.Version=${VERSION} -X github.com/0xERR0R/blocky/util.BuildTime=$(date -d "${BUILD_TIME}" '+%Y%m%d-%H%M%S') -X github.com/0xERR0R/blocky/util.Architecture=${GOARCH}"
```

For using `make` since they mentioned editing the Makefile I thought it wasn't actually a blocker, but doesn't matter either way.

**Reply by kwitsch** [reply]
Ahhh sorry missed removing the shell 🤦‍♂️

**Reply by tonho911** [reply]
This build was successful but when running ./blocky an error occurred
![Screenshot_109](https://github.com/0xERR0R/blocky/assets/148846306/97b495be-efb6-43cd-a5f9-f0337ecad62a)

**Reply by tonho911** [reply]
This build has an error when executing in the terminal, the error looks like this
![Uploading Screenshot_108.jpg…]()

**Reply by ThinkChaos** [reply]
Seems like the executable is not valid for some reason. Can you post the output of `file blocky` and `head blocky`?

**Reply by kwitsch** [reply]
I think I got something like that in the scratch images since there was a library missing.
The simplest way to fix it was using Zig as underlaying C compiler.

Sadly I've forgotten which library it was...🤔
